compiler/next/test/uast/testStringify.cpp

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright 2021-2022 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <climits>
+#include <map>
+#include <vector>
+#include "chpl/queries/Context.h"
+#include "chpl/queries/UniqueString.h"
+#include "chpl/uast/all-uast.h"
+#include "chpl/parsing/Parser.h"
+
+// always check assertions in this test
+#ifdef NDEBUG
+#undef NDEBUG
+#endif
+
+using namespace chpl;
+using namespace uast;
+using namespace parsing;
+
+
+// This macro just helps keep the tests a little cleaner
+// rather than redefine all this for every type we want to test
+#define TEST_STRINGIFY_TYPE(val)                          \
+    {                                                     \
+      std::ostringstream ss;                              \
+      stringify<std::decay_t<decltype(val)>> stringifier; \
+      stringifier(ss, DEBUG_DETAIL, val);                 \
+      assert(!ss.str().empty());                          \
+    }
+
+
+static void stringifyNode(const ASTNode* node, chpl::StringifyKind kind) {
+    // recurse through the nodes and make sure each can call stringify()
+    // and produce a non-empty result (for now)
+    // this is a little convoluted as each ASTNode is also calling
+    // ASTNode.dumpHelper() on all of its children
```

Comment:
This is fine, it's calling `dumpHelper` for now which is recursing, but when we get to actually writing AST dumpers we may want visitors to manually control the recursion (depends on how complicated/fine-grained we want the output to be). So it may not always be the case that `dumpHelper` is driving the tree walk.

One slight hole here is that calling `dumpHelper` still leaves room for a node to print nothing about itself when stringified, but it can still print a non-empty result because its children _do_ print something. Not sure if it's worth worrying about, though. I think just starting to add detail to the AST dumps is more important.

Reply:
I added a test for an empty module